### ISSUE TITLE
Can this let us simplify handling of void?

### ISSUE THREAD

**Issue by munificent** [chalin/DEP-non-null#2]
The spec's treatment of `void` is a bit peculiar. In most ways, void functions act like functions whose return type is Null, but there are a couple of places where it's treated specially.

I believe that's to ensure you get static warnings if you "do stuff" with the result of a void function call. For example, given:

``` dart
void returnsVoid() {}
```

The spec wants warnings here:

``` dart
Object obj = returnsVoid();
```

And here:

``` dart
returnsVoid().toString();
```

If `void` was just a syntax for writing a `Null` return type annotation, you wouldn't get these warnings because Object is a supertype of Null. But since your proposal reorganizes the root hierarchy, I believe we may be able to simplify the spec around.

We could just say that `void` as a return type is equivalent to `Null` and remove the idea of a "void type" completely. That would still give you the first warning above since Null isn't a subtype of Object anymore.

You'd also get the second warning _if_ we didn't move `toString()` up to `_NullOrObject`. Personally, though, I think it should be there, and this code shouldn't have a warning, but maybe that's just me.

What do you think?

@gbracha? Thoughts?

**Comment by chalin** [comment]
Right, this seems like an excellent opportunity! I refactored the `_Anything` class and added (f134164270d6c6d5f559ff1b911473ece5da0afc) its definition to the proposal (see [B.2.1](https://github.com/chalin/DEP-non-null/blob/master/doc/dep-non-null-AUTOGENERATED-DO-NOT-EDIT.md#new-root)):

``` dart
abstract class _Anything { const _Anything(); }

abstract class _Basic extends _Anything {
  int get hashCode;
  String toString();
  dynamic noSuchMethod(Invocation invocation);
  Type get runtimeType;
}

class Object extends _Anything implements _Basic {
  const Object();
  bool operator ==(other) => identical(this, other);
  ... // Methods of _Basic are all declared external
}
```

This organization allows us to write the following definition for `void`:

``` dart
abstract class void extends _Anything {}
class Null extends _Anything implements _Basic, void { /* Same as in DartC */ }
```

One could then make use of `void` as a type argument to futures as in `Future<void>`.

**Comment by chalin** [comment]
Also: I know that this is not possible currently, but it would be nice if we could say that `Null` was defined as an `enum` (since that is effectively what it is). I.e., something like:

``` java
enum Null /* extends _Anything implements $void, _Basic */ 
  { null }
```

**Comment by gmosx** [comment]
@munificent you propose:

Null returnsVoid() {}

??

 (kinda like it, never liked void ;-))

**Comment by munificent** [comment]
> One could then make use of void as a type argument to futures as in Future<void>.

We've already got `Future<Null>` for that case (though in practice everyone just does `Future`).

> @munificent you propose:
> 
> ```
> Null returnsVoid() {}
> ```

I'm proposing that `void` is purely syntactic sugar for a return type of `Null` and that the spec no longer talks about a "void type".
